File: data_structures/heap/go/heap.go
package datastructures

import "errors"

var ErrEmptyHeap = errors.New("heap is empty")

// MinHeap - array-based binary heap where parent <= children
type MinHeap struct {
	items []int
}

func NewMinHeap() *MinHeap {
	return &MinHeap{
		items: []int{},
	}
}

func (h *MinHeap) Insert(val int) {
	h.items = append(h.items, val)
	h.siftUp(len(h.items) - 1)
}

func (h *MinHeap) ExtractMin() (int, error) {
	if h.IsEmpty() {
		return 0, ErrEmptyHeap
	}

	min := h.items[0]
	last := len(h.items) - 1
	h.items[0] = h.items[last]
	h.items = h.items[:last]

	if !h.IsEmpty() {
		h.siftDown(0)
	}

	return min, nil
}

func (h *MinHeap) Peek() (int, error) {
	if h.IsEmpty() {
		return 0, ErrEmptyHeap
	}

	return h.items[0], nil
}

func (h *MinHeap) Size() int {
	return len(h.items)
}

func (h *MinHeap) IsEmpty() bool {
	return len(h.items) == 0
}

func (h *MinHeap) siftUp(index int) {
	for index > 0 {
		parent := (index - 1) / 2
		if h.items[index] < h.items[parent] {
			h.items[index], h.items[parent] = h.items[parent], h.items[index]
			index = parent
		} else {
			break
		}
	}
}

func (h *MinHeap) siftDown(index int) {
	size := len(h.items)
	for {
		smallest := index
		left := 2*index + 1
		right := 2*index + 2

		if left < size && h.items[left] < h.items[smallest] {
			smallest = left
		}
		if right < size && h.items[right] < h.items[smallest] {
			smallest = right
		}

		if smallest == index {
			break
		}

		h.items[index], h.items[smallest] = h.items[smallest], h.items[index]
		index = smallest
	}
}

// MaxHeap - array-based binary heap where parent >= children
type MaxHeap struct {
	items []int
}

func NewMaxHeap() *MaxHeap {
	return &MaxHeap{
		items: []int{},
	}
}

func (h *MaxHeap) Insert(val int) {
	h.items = append(h.items, val)
	h.siftUp(len(h.items) - 1)
}

func (h *MaxHeap) ExtractMax() (int, error) {
	if h.IsEmpty() {
		return 0, ErrEmptyHeap
	}

	max := h.items[0]
	last := len(h.items) - 1
	h.items[0] = h.items[last]
	h.items = h.items[:last]

	if !h.IsEmpty() {
		h.siftDown(0)
	}

	return max, nil
}

func (h *MaxHeap) Peek() (int, error) {
	if h.IsEmpty() {
		return 0, ErrEmptyHeap
	}

	return h.items[0], nil
}

func (h *MaxHeap) Size() int {
	return len(h.items)
}

func (h *MaxHeap) IsEmpty() bool {
	return len(h.items) == 0
}

func (h *MaxHeap) siftUp(index int) {
	for index > 0 {
		parent := (index - 1) / 2
		if h.items[index] > h.items[parent] {
			h.items[index], h.items[parent] = h.items[parent], h.items[index]
			index = parent
		} else {
			break
		}
	}
}

func (h *MaxHeap) siftDown(index int) {
	size := len(h.items)
	for {
		largest := index
		left := 2*index + 1
		right := 2*index + 2

		if left < size && h.items[left] > h.items[largest] {
			largest = left
		}
		if right < size && h.items[right] > h.items[largest] {
			largest = right
		}

		if largest == index {
			break
		}

		h.items[index], h.items[largest] = h.items[largest], h.items[index]
		index = largest
	}
}

// Heapify builds a MinHeap from an unsorted slice in O(n)
func Heapify(arr []int) *MinHeap {
	items := make([]int, len(arr))
	copy(items, arr)

	h := &MinHeap{items: items}

	for i := len(items)/2 - 1; i >= 0; i-- {
		h.siftDown(i)
	}

	return h
}

File: data_structures/heap/python/heap.py
class MinHeap:
    def __init__(self):
        self._items: list[int] = []

    def insert(self, val: int) -> None:
        self._items.append(val)
        self._sift_up(len(self._items) - 1)

    def extract_min(self) -> int:
        if self.is_empty():
            raise IndexError("heap is empty")
        min_val = self._items[0]
        last = self._items.pop()
        if self._items:
            self._items[0] = last
            self._sift_down(0)
        return min_val

    def peek(self) -> int:
        if self.is_empty():
            raise IndexError("heap is empty")
        return self._items[0]

    def size(self) -> int:
        return len(self._items)

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def _sift_up(self, index: int) -> None:
        while index > 0:
            parent = (index - 1) // 2
            if self._items[index] < self._items[parent]:
                self._items[index], self._items[parent] = self._items[parent], self._items[index]
                index = parent
            else:
                break

    def _sift_down(self, index: int) -> None:
        size = len(self._items)
        while True:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2
            if left < size and self._items[left] < self._items[smallest]:
                smallest = left
            if right < size and self._items[right] < self._items[smallest]:
                smallest = right
            if smallest == index:
                break
            self._items[index], self._items[smallest] = self._items[smallest], self._items[index]
            index = smallest


class MaxHeap:
    def __init__(self):
        self._items: list[int] = []

    def insert(self, val: int) -> None:
        self._items.append(val)
        self._sift_up(len(self._items) - 1)

    def extract_max(self) -> int:
        if self.is_empty():
            raise IndexError("heap is empty")
        max_val = self._items[0]
        last = self._items.pop()
        if self._items:
            self._items[0] = last
            self._sift_down(0)
        return max_val

    def peek(self) -> int:
        if self.is_empty():
            raise IndexError("heap is empty")
        return self._items[0]

    def size(self) -> int:
        return len(self._items)

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def _sift_up(self, index: int) -> None:
        while index > 0:
            parent = (index - 1) // 2
            if self._items[index] > self._items[parent]:
                self._items[index], self._items[parent] = self._items[parent], self._items[index]
                index = parent
            else:
                break

    def _sift_down(self, index: int) -> None:
        size = len(self._items)
        while True:
            largest = index
            left = 2 * index + 1
            right = 2 * index + 2
            if left < size and self._items[left] > self._items[largest]:
                largest = left
            if right < size and self._items[right] > self._items[largest]:
                largest = right
            if largest == index:
                break
            self._items[index], self._items[largest] = self._items[largest], self._items[index]
            index = largest


def heapify(arr: list[int]) -> MinHeap:
    h = MinHeap()
    h._items = arr.copy()
    for i in range(len(h._items) // 2 - 1, -1, -1):
        h._sift_down(i)
    return h
